zeppelin-web/src/app/notebook/notebook.controller.js

```diff
@@ -1069,8 +1064,55 @@ function NotebookCtrl ($scope, $route, $routeParams, $location, $rootScope,
     $scope.note.config.personalizedMode = isPersonalized
   })
 
+  $scope.$on('$routeChangeStart', function (event, next, current) {
+    if ($scope.note && $scope.note.paragraphs) {
+      _.forEach($scope.note.paragraphs, function (par) {
+        if ($scope.allowLeave === true) {
+          return
+        }
+        let thisScope = angular.element(
+          '#' + par.id + '_paragraphColumn_main').scope()
+
+        if (thisScope.dirtyText === undefined ||
+          thisScope.originalText === undefined ||
+          thisScope.dirtyText === thisScope.originalText) {
+          return true
+        } else {
+          event.preventDefault()
+
+          BootstrapDialog.show({
```

Comment:
Can we extract `BootstrapDialog.show` outside of the for loop? 

It's actually triggered once. But hope it looks like can be triggered multiple times as it's in the for loop